extensions/smallrye-reactive-messaging/runtime/src/main/java/io/quarkus/smallrye/reactivemessaging/runtime/QuarkusWorkerPoolRegistry.java

```diff
@@ -56,6 +114,24 @@ public <T> Uni<T> executeWork(Context currentContext, Uni<T> uni, String workerN
                 return currentContext.executeBlocking(Uni.createFrom().deferred(() -> uni), ordered);
             }
             return executionHolder.vertx().executeBlocking(uni, ordered);
+        } else if (virtualThreadConcurrency.containsKey(workerName)) {
+            Semaphore semaphore = virtualThreadConcurrency.get(workerName);
+            boolean acquired = semaphore.tryAcquire();
```

Comment:
@ozangunalp as discussed yesterday, you confirm that the number of permits is what drive the requests. 

(CC @jponge - we use `merge(concurrency)` to control the flow (and also avoid OOM))

Reply:
Yes, this is just a draft. The idea is to use requests controlled with the `merge(concurrency)` config in reactive messaging to adjust the max concurrency for virtual threads. So the semaphores will disappear.